File: music-android/app/src/main/java/com/sismics/music/adapter/AlbumAdapter.java
package com.sismics.music.adapter;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.androidquery.AQuery;
import com.androidquery.callback.BitmapAjaxCallback;
import com.sismics.music.R;
import com.sismics.music.event.TrackCacheStatusChangedEvent;
import com.sismics.music.model.FullAlbum;
import com.sismics.music.util.CacheUtil;
import com.sismics.music.util.PreferenceUtil;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.List;

/**
 * Adapter for albums list.
 * 
 * @author bgamard
 */
public class AlbumAdapter extends BaseAdapter implements Filterable {

    private AQuery aq;
    private Activity activity;
    private List<FullAlbum> onlineAlbums; // TODO Merge those 2 lists in one
    private List<FullAlbum> cachedAlbums;
    private List<FullAlbum> displayedAlbums;
    private String authToken;
    private String serverUrl;
    private boolean offlineMode;

    /**
     * Constructor.
     * @param activity Context activity
     */
    public AlbumAdapter(Activity activity, List<FullAlbum> onlineAlbums, List<FullAlbum> cachedAlbums, boolean offlineMode) {
        this.activity = activity;
        this.offlineMode = offlineMode;
        this.aq = new AQuery(activity);
        this.authToken = PreferenceUtil.getAuthToken(activity);
        this.serverUrl = PreferenceUtil.getServerUrl(activity);
        this.onlineAlbums = onlineAlbums;
        this.cachedAlbums = cachedAlbums;
        resetDisplayedAlbums();
    }

    @Override
    public View getView(int position, View view, ViewGroup parent) {
        ViewHolder holder;
        
        if (view == null) {
            LayoutInflater vi = (LayoutInflater) activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = vi.inflate(R.layout.list_item_album, parent, false);
            aq.recycle(view);
            holder = new ViewHolder();
            holder.albumName = aq.id(R.id.albumName).getTextView();
            holder.artistName = aq.id(R.id.artistName).getTextView();
            holder.imgCover = aq.id(R.id.imgCover).getImageView();
            holder.cached = aq.id(R.id.cached).getImageView();
            holder.overflow = aq.id(R.id.overflow).getView();
            view.setTag(holder);
        } else {
            aq.recycle(view);
            holder = (ViewHolder) view.getTag();
        }

        FullAlbum album = getItem(position);

        // Album cover
        String coverUrl = serverUrl + "/api/album/" + album.getAlbum().getId() + "/albumart/small";
        aq.id(holder.imgCover).image(new BitmapAjaxCallback()
                .url(coverUrl)
                .animation(AQuery.FADE_IN_NETWORK)
                .cookie("auth_token", authToken)
        );

        // Filling album data
        holder.albumName.setText(album.getAlbum().getName());
        holder.artistName.setText(album.getArtist().getName());
        final View cached = holder.cached;
        cached.setVisibility(CacheUtil.isAlbumCached(activity, album.getAlbum().getId()) ? View.VISIBLE : View.GONE);

        // Configuring popup menu
        aq.id(holder.overflow).clicked(v -> {
            PopupMenu popup = new PopupMenu(activity, v);
            popup.inflate(R.menu.list_item_album);

            // Menu actions
            popup.setOnMenuItemClickListener(item -> {
                CacheUtil.removeAlbum(activity, album.getArtist().getId(), album.getAlbum().getId());
                EventBus.getDefault().post(new TrackCacheStatusChangedEvent(null));
                return true;
            });

            popup.show();
        });

        return view;
    }

    @Override
    public int getCount() {
        return displayedAlbums.size();
    }

    @Override
    public FullAlbum getItem(int position) {
        return displayedAlbums.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    public void resetOnlineAlbums() {
        onlineAlbums = new ArrayList<>();
    }

    public void addOnlineAlbums(List<FullAlbum> albums) {
        onlineAlbums.addAll(albums);
        resetDisplayedAlbums();
        notifyDataSetChanged();
    }

    public void setCachedAlbums(List<FullAlbum> albums) {
        cachedAlbums = albums;
        resetDisplayedAlbums();
        notifyDataSetChanged();
    }

    private void resetDisplayedAlbums() {
        displayedAlbums = offlineMode ? cachedAlbums : onlineAlbums;
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                if (!offlineMode) {
                    return null;
                }
                FilterResults results = new FilterResults();
                if (constraint == null || constraint.length() == 0) {
                    results.values = cachedAlbums;
                    results.count = cachedAlbums.size();
                    return results;
                }

                // Search in album name and artist name
                List<FullAlbum> values = new ArrayList<>();
                String filter = constraint.toString().toLowerCase();
                for (FullAlbum album : cachedAlbums) {
                    if (album.getAlbum().getName().toLowerCase().contains(filter)
                            || album.getArtist().getName().toLowerCase().contains(filter)) {
                        values.add(album);
                    }
                }
                results.values = values;
                results.count = values.size();
                return results;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                if (results != null) {
                    displayedAlbums = (List<FullAlbum>) results.values;
                    notifyDataSetChanged();
                }
            }
        };
    }

    public void setOfflineMode(boolean offlineMode) {
        this.offlineMode = offlineMode;
        resetDisplayedAlbums();
        notifyDataSetChanged();
    }

    /**
     * Album ViewHolder.
     * 
     * @author bgamard
     */
    private static class ViewHolder {
        TextView albumName;
        TextView artistName;
        ImageView imgCover;
        ImageView cached;
        View overflow;
    }
}

File: music-android/app/src/main/java/com/sismics/music/adapter/AlbumAdapter.kt
package com.sismics.music.adapter

import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.androidquery.AQuery
import com.androidquery.callback.BitmapAjaxCallback
import com.sismics.music.R
import com.sismics.music.event.TrackCacheStatusChangedEvent
import com.sismics.music.util.CacheUtil
import com.sismics.music.util.PreferenceUtil
import de.greenrobot.event.EventBus
import org.json.JSONArray
import org.json.JSONObject

/**
 * Adapter for albums list.
 *
 * @author bgamard
 */
class AlbumAdapter(
        private val activity: Activity,
        private var originalAlbums: JSONArray?,
        private val cachedAlbumSet: Set<String>,
        private var offlineMode: Boolean) : BaseAdapter(), Filterable {

    private val aq: AQuery
    private var allAlbums: JSONArray? = null
    private var albums: JSONArray? = null
    private val authToken: String?
    private val serverUrl: String?

    init {
        this.aq = AQuery(activity)
        this.authToken = PreferenceUtil.getAuthToken(activity)
        this.serverUrl = PreferenceUtil.getServerUrl(activity)
        computeAlbumList()
        this.albums = this.allAlbums
    }

    private fun computeAlbumList() {
        if (offlineMode) {
            allAlbums = JSONArray()
            for (i in 0..originalAlbums!!.length() - 1) {
                val album = originalAlbums!!.optJSONObject(i)
                if (cachedAlbumSet.contains(album.optString("id"))) {
                    allAlbums!!.put(album)
                }
            }
        } else {
            allAlbums = originalAlbums
        }
    }

    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        var view = view
        val holder: ViewHolder

        if (view == null) {
            val vi = activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            view = vi.inflate(R.layout.list_item_album, null)
            aq.recycle(view)
            holder = ViewHolder()
            holder.albumName = aq.id(R.id.albumName).textView
            holder.artistName = aq.id(R.id.artistName).textView
            holder.imgCover = aq.id(R.id.imgCover).imageView
            holder.cached = aq.id(R.id.cached).imageView
            holder.overflow = aq.id(R.id.overflow).view
            view!!.tag = holder
        } else {
            aq.recycle(view)
            holder = view.tag as ViewHolder
        }

        val album = getItem(position)

        // Album cover
        val albumId = album.optString("id")
        val coverUrl = "$serverUrl/api/album/$albumId/albumart/small"
        if (aq.shouldDelay(position, view, parent, coverUrl)) {
            aq.id(holder.imgCover).image(null as Bitmap?)
        } else {
            aq.id(holder.imgCover).image(BitmapAjaxCallback().url(coverUrl).animation(AQuery.FADE_IN_NETWORK).cookie("auth_token", authToken))
        }

        // Filling album data
        holder.albumName!!.text = album.optString("name")
        val artist = album.optJSONObject("artist")
        holder.artistName!!.text = artist.optString("name")
        val cached = holder.cached
        cached!!.visibility = if (cachedAlbumSet.contains(albumId)) View.VISIBLE else View.GONE

        // Configuring popup menu
        aq.id(holder.overflow).clicked { v ->
            val popup = PopupMenu(activity, v)
            popup.inflate(R.menu.list_item_album)

            // Menu actions
            popup.setOnMenuItemClickListener {
                CacheUtil.removeAlbum(albumId)
                EventBus.getDefault().post(TrackCacheStatusChangedEvent(null))
                true
            }

            popup.show()
        }

        return view
    }

    override fun getCount(): Int {
        return albums!!.length()
    }

    override fun getItem(position: Int): JSONObject {
        return albums!!.optJSONObject(position)
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    fun setAlbums(albums: JSONArray) {
        this.originalAlbums = albums
        computeAlbumList()
        this.albums = allAlbums
        notifyDataSetChanged()
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence?): Filter.FilterResults {
                val results = Filter.FilterResults()
                if (constraint == null || constraint.isEmpty()) {
                    results.values = allAlbums
                    results.count = allAlbums!!.length()
                    return results
                }

                // Search in album name and artist name
                val values = JSONArray()
                val filter = constraint.toString().toLowerCase()
                for (i in 0..allAlbums!!.length() - 1) {
                    val album = allAlbums!!.optJSONObject(i)
                    if (album.optString("name").toLowerCase().contains(filter) || album.optJSONObject("artist").optString("name").toLowerCase().contains(filter)) {
                        values.put(album)
                    }
                }
                results.values = values
                results.count = values.length()
                return results
            }

            override fun publishResults(constraint: CharSequence, results: Filter.FilterResults) {
                albums = results.values as JSONArray
                notifyDataSetChanged()
            }
        }
    }

    fun setOfflineMode(offlineMode: Boolean) {
        this.offlineMode = offlineMode
        computeAlbumList()
        notifyDataSetChanged()
    }

    /**
     * Album ViewHolder.
     * @author bgamard
     */
    private class ViewHolder {
        internal var albumName: TextView? = null
        internal var artistName: TextView? = null
        internal var imgCover: ImageView? = null
        internal var cached: ImageView? = null
        internal var overflow: View? = null
    }
}
